service/awsconfig/v10/resource/s3bucket/create.go

```diff
@@ -16,55 +17,88 @@ func (r *Resource) ApplyCreateChange(ctx context.Context, obj, createChange inte
 
-		_, err = r.clients.S3.CreateBucket(&s3.CreateBucketInput{
-			Bucket: aws.String(bucketInput.Name),
-		})
-		if IsBucketAlreadyExists(err) || IsBucketAlreadyOwnedByYou(err) {
-			// Fall through.
-			return nil
-		}
-		if err != nil {
-			return microerror.Mask(err)
-		}
+			_, err = r.clients.S3.CreateBucket(&s3.CreateBucketInput{
+				Bucket: aws.String(bucketInput.Name),
+			})
+			if IsBucketAlreadyExists(err) || IsBucketAlreadyOwnedByYou(err) {
+				// Fall through.
+				return nil
+			}
+			if err != nil {
+				return microerror.Mask(err)
+			}
 
-		_, err := r.clients.S3.PutBucketTagging(&s3.PutBucketTaggingInput{
-			Bucket: aws.String(bucketInput.Name),
-			Tagging: &s3.Tagging{
-				TagSet: r.getS3BucketTags(customObject),
-			},
-		})
-		if err != nil {
-			return microerror.Mask(err)
-		}
+			_, err = r.clients.S3.PutBucketTagging(&s3.PutBucketTaggingInput{
+				Bucket: aws.String(bucketInput.Name),
+				Tagging: &s3.Tagging{
+					TagSet: r.getS3BucketTags(customObject),
+				},
+			})
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			if bucketInput.IsDeliveryLog {
+				_, err = r.clients.S3.PutBucketAcl(&s3.PutBucketAclInput{
+					Bucket:       aws.String(key.TargetLogBucketName(customObject)),
+					GrantReadACP: aws.String(key.LogDeliveryURI),
+					GrantWrite:   aws.String(key.LogDeliveryURI),
+				})
+				if err != nil {
+					return microerror.Mask(err)
+				}
+			}
+
+			if bucketInput.LoggingEnabled {
+				_, err = r.clients.S3.PutBucketLogging(&s3.PutBucketLoggingInput{
+					Bucket: aws.String(bucketInput.Name),
+					BucketLoggingStatus: &s3.BucketLoggingStatus{
+						LoggingEnabled: &s3.LoggingEnabled{
+							TargetBucket: aws.String(key.TargetLogBucketName(customObject)),
+							TargetPrefix: aws.String(bucketInput.Name),
+						},
+					},
+				})
+				if err != nil {
+					return microerror.Mask(err)
+				}
+			}
 
-		r.logger.LogCtx(ctx, "level", "debug", "message", "creating S3 bucket: created")
-	} else {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "creating S3 bucket: already created")
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating S3 bucket %q: created", bucketInput.Name))
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating S3 bucket %q: already created", bucketInput.Name))
+		}
 	}
+
 	return nil
 }
 
 func (r *Resource) newCreateChange(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error) {
-	currentBucket, err := toBucketState(currentState)
+	currentBuckets, err := toBucketState(currentState)
 	if err != nil {
 		return nil, microerror.Mask(err)
 	}
-	desiredBucket, err := toBucketState(desiredState)
+	desiredBuckets, err := toBucketState(desiredState)
 	if err != nil {
 		return nil, microerror.Mask(err)
 	}
 
-	if currentBucket.Name == "" || desiredBucket.Name != currentBucket.Name {
-		return desiredBucket, nil
+	createState := []BucketState{}
```

Comment:
I have just followed the same approach I saw for others resource, and for example [here](https://github.com/giantswarm/aws-operator/blob/master/service/awsconfig/v9/resource/s3bucket/create.go#L69) we return the empty object instead of `nil`. Does it change when we manage slices? why? and if so how it would look like?

Reply:
This is because of the difference between values and pointers. The point is the zero value. The zero value of a value is the initialized empty value. The zero value of a pointer is nil. A slice is a pointer.

Reply:
ok nice!

But is this will work too?
`var bs []BucketState`
If you don't set any value is `nil` by default, right? or is it better be more explicit?

Reply:
`var bs []BucketState` yes, that works, and it sets the value to `nil`. In examples above I tried to be as clear (explicit) as possible.